src/Utils/parameters.jl

```diff
@@ -0,0 +1,343 @@
+Base.setindex!(ps::SpeedyParams, x, nm::Symbol) = setindex!(ps, x, :, nm)
+@inline function Base.setindex!(ps::SpeedyParams, x, ::Colon, nm::Symbol)
+    # include fieldname only (no component)
+    if nm == :idx
+        error("cannot set :idx")
+    elseif nm == :fieldname
+        error("cannot set :fieldname")
+    else
+        newparent = if nm in keys(ps)
+            ModelParameters._setindex(ModelParameters.parent(ps), Tuple(x), nm)
+        else
+            ModelParameters._addindex(ModelParameters.parent(ps), Tuple(x), nm)
+        end
+        return SpeedyParams(newparent)
+    end
+end
+
+Base.keys(params::SpeedyParams) = (:idx, :fieldname, keys(first(params))...)
+
+function Base.vec(params::SpeedyParams)
+    # recursively unpack params and build component vector
+    return ComponentArray(unpack_params(params))
+end
+
+# Override internal ModelParameters method _columntypes to condense type names in table schema (just to look nicer)
+# TODO: propose this change upstream in ModelParameters.jl
+ModelParameters._columntypes(ps::SpeedyParams) = map(k -> promote_type(map(typeof, getindex(ps, k))...), keys(ps)) 
+
+# parameters method interface
+"""
+    $SIGNATURES
+
+Returns a "parameterized" copy of the given data structure. For numeric types, this simply wraps
+the given value in a `Param`-like type. Dispatches can be added for compound data types that
+reconstruct the data type with `Param`-like wrappers on all relevant parameters.
+"""
+parameters(obj; kwargs...) = parameters(SpeedyParam, obj; kwargs...)
+parameters(::Type{PT}, obj; kwargs...) where {PT<:AbstractParam} = (;)
+parameters(::Type{PT}, param::AbstractParam; kwargs...) where {PT<:AbstractParam} = PT(merge(parent(param), kwargs))
+parameters(::Type{PT}, x::Union{Number,AbstractArray}; kwargs...) where {PT<:AbstractParam} = PT(x; kwargs...)
+
+"""
+    $SIGNATURES
+
+Convenience method that creates a model parameter from its property with the given `name` and optional extra attributes in `kwargs`.
+A parameter attribute `copmonenttype` is automatically added with value `T`.
+"""
+parameterof(obj::T, ::Val{propname}; kwargs...) where {T,propname} = parameterof(SpeedyParam, obj, Val{propname}(); kwargs...)
+parameterof(::Type{PT}, obj::T, ::Val{propname}; kwargs...) where {PT,T,propname} = parameters(PT, getproperty(obj, propname); componenttype=T, kwargs...)
+
+# reconstruct
+
+"""
+    $SIGNATURES
+
+Reconstructs the given data structure with the given `values`. If `values` is a `NamedTuple`, the structure
+must match that of `obj`. This function is used to reconstruct model types from `SpeedyParams`.
+"""
+reconstruct(obj::T, value::T) where {T} = value
+reconstruct(obj::AbstractParam, value::T) where {T} = ModelParameters.update(obj, Tuple(value))
+reconstruct(obj::NamedTuple{keys,V}, values::NamedTuple{keys,V}) where {keys,V<:Tuple} = values
+reconstruct(obj, values::ComponentArray) = reconstruct(obj, NamedTuple(values))
```

Comment:
Just wondering, if we generally keep the parameters in a ComponentArray in ML tasks, maybe it would be better to have a routine that directly reconstructs from a ComponentArray as well. Or is `NamedTuple(::ComponentArray)` trivial and non-allocating?